### DISCUSSION TITLE
Multiple `useRequestPreSend`: Should async event callbacks be awaited?

### DISCUSSION THREAD

**Discussion by leomelzer** [rpldy/react-uploady#662] [Q&A] [answered]
Hello!

Thanks for making and maintaining react-uploady, it's a joy to use! 

I implemented signing with S3 parameters using `useRequestPreSend` and wanted to add client-side image compression to my pipeline. For that I simply created another `useRequestPreSend` before the signing one returning my altered `items`. 

However it turned out my compressed images weren't being used for uploading...after some digging it turns out the registered events in `life-events` are in fact executed (or rather kicked-off) in sequence _but_ the result not awaited.

Relevant code: https://github.com/rpldy/react-uploady/blob/master/packages/core/life-events/src/lifeEvents.js#L142-L166

I would have expected something like

```ts
results = regs.map(async (r: RegItem): any => {
            let result;

            if (r.once) {
                removeRegItem(this, name, r.cb);
			}

			if (packValue) {
                result = r.cb(...packValue);
            } else if (!args.length) {
				result = r.cb();
			} else {
				result = r.cb(...args);
			}

			return await result;
		})
			.filter((result: any): any => !isUndefined(result))
			.map((result: any): Promise<any> => isPromise(result) ? result : Promise.resolve(result));
```

I can see both sides so unsure whether this is in fact a bug or rather a conscious design decision. 

For now I'll roll my code into one `useRequestPreSend` which is completely fine. I just figured I share this for future readers.

Thank you!

**Comment by yoavniran** [comment, marked as answer]
thanks @leomelzer 
Indeed this was a design decision regarding how to handle pub/sub within Uploady.

And as you mentioned, a single callback to handle these two aspects is the recommended way.
